bson/bsonrw/copier.go

```diff
@@ -61,9 +77,19 @@ func (c Copier) CopyDocumentFromBytes(dst ValueWriter, src []byte) error {
 	return dw.WriteDocumentEnd()
 }
 
+// CopyBytesToArrayWriter copies the values from a BSON Array represented as a []byte to an
+// ArrayWriter.
+func (c Copier) CopyBytesToArrayWriter(dst ArrayWriter, src []byte) error {
+	return c.copyBytesToValueWriter(dst.(ValueWriter), src, true)
```

Comment:
I like the `copyByteToValueWriter` method to abstract out the common logic to copy arrays and documents, but the cast to `ValueWriter` here and in `CopyBytesToDocumentWriter` isn't technically correct: it works for our defaults because the unexported [bsonrw.valueWriter](https://github.com/mongodb/mongo-go-driver/blob/54681e1af6f3c0ea4761ae8a998034bcbd4e7a43/bson/bsonrw/value_writer.go#L139) type implements `ValueWriter`, `ArrayWriter`, and `DocumentWriter` in a single type, but these interfaces are all public, so a user could implement them as separate types and manually call these methods, which would probably cause a panic.

Here's a possible proposal (not fully fleshed out, so let's brainstorm how to clean it up if you like it):

```
type valueWriterGetterFn func(key string) ValueWriter

func CopyBytesToDocumentWriter(dst DocumentWriter, src []byte) error {
    vwGetter := func(key string) (ValueWriter, error) {
        return dst.WriteDocumentElement(key)
    }

    return c.copyBytestoDocumentWriter(src, vwGetter)
}

// CopyBytesToArrayWriter is similar to CopyBytesToDocumentWriter

func copyBytesToDocumentWriter(src []byte, vwGetterFn valueWriterGetterFn) {
    // replace the WriteDocumentElement/WriteArrayElement calls with vwGetterFn()
}
```

Reply:
Good catch, and thanks again for the small lesson on type-casting offline. I've used something very similar to your code above with a new `writeElementFn` type.